File: scala/test/challenges/cracking/linkedLists/NodeTest.scala
package challenges.cracking.linkedLists

import org.scalatest.{BeforeAndAfter, FunSuite}

class NodeTest extends FunSuite with BeforeAndAfter {

  var n1:Node = _

  before {
    n1 = new Node(1)

    n1.appendToTail(3)
    n1.appendToTail(2)
    n1.appendToTail(2)
    n1.appendToTail(4)
    n1.appendToTail(5)

  }

  test("new list have 6 nodes") {
    assert(n1.size == 6)
  }

  test("delete data 3 node") {
    n1.deleteNode(3)
    assert(n1.size == 5)
    assert(n1.next.get.data != 3)
  }

  test("removing list duplicates") {
    n1.removeDuplicates()
    assert(n1.size == 5)
  }

  test("get last 2 nodes") {
    val nth:Option[Node] = n1.nthToLast(2)
    assert(nth.nonEmpty)
    assert(nth.get.size == 2)
  }

  test("get last 20 nodes") {
    val nth:Option[Node] = n1.nthToLast(20)
    assert(nth.isEmpty)
  }

  test("delete first node") {
    assert(n1.delete())
    assert(n1.size == 5)
    assert(n1.data != 1)
  }

  test("delete last node") {
    val lastNode: Node = n1.nthToLast(1).get
    assert(!lastNode.delete())
  }

  test("sum lists") {
    val lastNode: Node = n1.nthToLast(3).get
    // 542231 + 542 = 542773
    assert(542773 == LinkedListUtil.sumLists(n1, lastNode))
  }

}

File: ruby/test/challenges/cracking/linkedLists/NodeTest.rb
require 'test-unit'
require '../../../../src/challenges/cracking/linkedLists/Node'


class NodeTest < Test::Unit::TestCase

  def test_all

    n1 = Node.new(1)
    n1.append_to_tail 3
    n1.append_to_tail 2
    n1.append_to_tail 2
    n1.append_to_tail 4
    n1.append_to_tail 5

    assert_equal(6, n1.size)

    n1.delete_node 3 # 1-2-2-4-5

    assert_not_equal(n1.next.data, 3)
    assert_equal(5, n1.size)

    n1.remove_duplicates() # 1-2-4-5

    assert_equal(4, n1.size)

    last_two = n1.nth_to_last(2) # 4-5
    assert_equal(2, last_two.size)
    assert_equal(last_two.data, 4)
    assert_equal(last_two.next.data, 5)

    n1.delete # 2-4-5

    assert_equal(3, n1.size)
    assert_equal(2, n1.data)

  end

end